core/src/avm2/activation.rs

```diff
@@ -0,0 +1,1627 @@
+    /// If this frame is already executing, that is an error condition.
+    pub fn lock(&mut self) -> Result<(), Error> {
+        if self.is_executing {
+            return Err("Attempted to execute the same frame twice".into());
+        }
+
+        self.is_executing = true;
+
+        Ok(())
+    }
+
+    /// Unlock the activation object. This allows future execution to run on it
+    /// again.
+    pub fn unlock_execution(&mut self) {
+        self.is_executing = false;
+    }
+
+    /// Retrieve a local register.
+    pub fn local_register(&self, id: u32) -> Result<Value<'gc>, Error> {
+        self.local_registers
+            .read()
+            .get(id)
+            .cloned()
+            .ok_or_else(|| format!("Out of bounds register read: {}", id).into())
+    }
+
+    /// Get the current scope stack.
+    pub fn scope(&self) -> Option<GcCell<'gc, Scope<'gc>>> {
+        self.scope
+    }
+
+    /// Set a new scope stack.
+    pub fn set_scope(&mut self, new_scope: Option<GcCell<'gc, Scope<'gc>>>) {
+        self.scope = new_scope;
+    }
+
+    /// Set a local register.
+    ///
+    /// Returns `true` if the set was successful; `false` otherwise
+    pub fn set_local_register(
+        &mut self,
+        id: u32,
+        value: impl Into<Value<'gc>>,
+        mc: MutationContext<'gc, '_>,
+    ) -> Result<(), Error> {
+        if let Some(r) = self.local_registers.write(mc).get_mut(id) {
+            *r = value.into();
+
+            Ok(())
+        } else {
+            Err(format!("Out of bounds register write: {}", id).into())
+        }
+    }
+
+    pub fn avm2(&mut self) -> &mut Avm2<'gc> {
+        self.avm2
+    }
+
+    /// Set the return value.
+    pub fn set_return_value(&mut self, value: Value<'gc>) {
+        self.return_value = Some(value);
+    }
+
+    /// Get the base prototype of the object that the currently executing
+    /// method was retrieved from, if one exists.
+    pub fn base_proto(&self) -> Option<Object<'gc>> {
+        self.base_proto
+    }
+
+    /// Retrieve a int from the current constant pool.
+    fn pool_int(
+        &self,
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        index: Index<i32>,
+    ) -> Result<i32, Error> {
+        value::abc_int(&method.abc(), index)
+    }
+
+    /// Retrieve a int from the current constant pool.
+    fn pool_uint(
+        &self,
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        index: Index<u32>,
+    ) -> Result<u32, Error> {
+        value::abc_uint(&method.abc(), index)
+    }
+
+    /// Retrieve a double from the current constant pool.
+    fn pool_double(
+        &self,
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        index: Index<f64>,
+    ) -> Result<f64, Error> {
+        value::abc_double(&method.abc(), index)
+    }
+
+    /// Retrieve a string from the current constant pool.
+    fn pool_string(
+        &self,
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        index: Index<String>,
+    ) -> Result<String, Error> {
```

Comment:
Can this be `Result<&str, Error>`?

Reply:
There isn't a suitable lifetime for the returned string; the AbcFile that the string comes from is owned by the `Gc<'gc, BytecodeMethod<'gc>>` and we can only get temporary borrows for it.

Reply:
I have not looked over everything thoroughly, but you can use `std::cell::Ref::map`, e.g. starting in `TranslationUnit`:
```
    pub fn abc_ref(&self) -> std::cell::Ref<AbcFile> {
        std::cell::Ref::map(self.0.read(), |tx| &*tx.abc.0)
    }
```
and then similarly in `BytecodeMethod`:
```
    pub fn abc_ref(&self) -> std::cell::Ref<AbcFile> {
        self.txunit.abc_ref()
    }
```
and so forth, `Activation::pool_string` takes `&BytecodeMethod` instead of a `Gc`. Not sure if this is a helpful design, but at least its possible in theory.

Reply:
That actually worked! I had to do some value calisthenics to get around the fact that there is no `Ref::map_result`, though. This code looks super verbose now:

```
let mut maybe_error = None;
let maybe_ref = Ref::map(method.abc_ref(), |abc| {
    match value::abc_string(abc, index) {
        Ok(s) => s,
        Err(e) => {
            maybe_error = Some(e);
            ""
        }
    }
});

if let Some(err) = maybe_error {
    Err(err)
} else {
    Ok(maybe_ref)
}
```

Reply:
Thanks to Dinnerbone's `AvmString` work this approach is actually obsolete anyway, so I'm going to mark this as resolved.